File: platform/osmedeus-dashboard/lib/api/orgs.ts
import { http } from "./http";
import { API_PREFIX } from "@/lib/api/prefix";
import { isDemoMode } from "./demo-mode";
import { mockOrgs } from "@/lib/mock/data/orgs";
import type {
  Org,
  OrgStats,
  CreateOrgInput,
  UpdateOrgInput,
  OrgAssignResult,
} from "@/lib/types/org";

function mapStats(raw: any, fallbackUUID: string, fallbackName: string): OrgStats {
  return {
    org_uuid: String(raw?.org_uuid ?? fallbackUUID),
    org_name: String(raw?.org_name ?? fallbackName),
    total_workspaces: Number(raw?.total_workspaces ?? 0) || 0,
    total_assets: Number(raw?.total_assets ?? 0) || 0,
    total_vulns: Number(raw?.total_vulns ?? 0) || 0,
    total_runs: Number(raw?.total_runs ?? 0) || 0,
    workspaces: Array.isArray(raw?.workspaces) ? raw.workspaces.map(String) : [],
  };
}

function mapOrg(raw: any): Org {
  const uuid = String(raw?.uuid ?? "");
  const name = String(raw?.name ?? "");
  return {
    uuid,
    name,
    description: typeof raw?.description === "string" ? raw.description : undefined,
    tags: Array.isArray(raw?.tags) ? raw.tags.map(String) : [],
    is_default: Boolean(raw?.is_default),
    created_at: String(raw?.created_at ?? ""),
    updated_at: String(raw?.updated_at ?? ""),
    stats: raw?.stats ? mapStats(raw.stats, uuid, name) : undefined,
  };
}

/**
 * List every org with its workspace/asset/vulnerability counts.
 *
 * The counts come embedded in this response, so there is no reason to also call
 * fetchOrgStats when rendering a list.
 */
export async function fetchOrgs(): Promise<Org[]> {
  if (isDemoMode()) {
    return mockOrgs.map((o) => ({ ...o }));
  }
  const res = await http.get(`${API_PREFIX}/orgs`);
  const payload = (res.data || {}) as any;
  const list = Array.isArray(payload.data) ? payload.data : [];
  return list.map(mapOrg);
}

/** Fetch one org by name or UUID. */
export async function fetchOrg(ref: string): Promise<Org> {
  if (isDemoMode()) {
    const found = mockOrgs.find((o) => o.uuid === ref || o.name === ref);
    if (!found) throw new Error(`404:org not found: ${ref}`);
    return { ...found };
  }
  const res = await http.get(`${API_PREFIX}/orgs/${encodeURIComponent(ref)}`);
  return mapOrg((res.data || {}).data);
}

/** Fetch aggregate counts for one org. */
export async function fetchOrgStats(ref: string): Promise<OrgStats> {
  if (isDemoMode()) {
    const found = mockOrgs.find((o) => o.uuid === ref || o.name === ref);
    if (!found?.stats) throw new Error(`404:org not found: ${ref}`);
    return { ...found.stats };
  }
  const res = await http.get(`${API_PREFIX}/orgs/${encodeURIComponent(ref)}/stats`);
  const payload = (res.data || {}) as any;
  return mapStats(payload.data, ref, ref);
}

export async function createOrg(input: CreateOrgInput): Promise<Org> {
  if (isDemoMode()) {
    throw new Error("400:Creating orgs is disabled in demo mode");
  }
  const res = await http.post(`${API_PREFIX}/orgs`, input);
  return mapOrg((res.data || {}).data);
}

export interface UpdateOrgResult {
  org: Org;
  assigned?: OrgAssignResult;
}

/**
 * Update an org's metadata and optionally assign workspaces to it.
 *
 * Assigning cascades the org stamp to every asset, vulnerability and run in
 * those workspaces — this is how pre-existing data gets grouped without
 * re-scanning.
 */
export async function updateOrg(ref: string, input: UpdateOrgInput): Promise<UpdateOrgResult> {
  if (isDemoMode()) {
    throw new Error("400:Editing orgs is disabled in demo mode");
  }
  const res = await http.put(`${API_PREFIX}/orgs/${encodeURIComponent(ref)}`, input);
  const payload = (res.data || {}) as any;
  return { org: mapOrg(payload.data), assigned: payload.assigned };
}

/** Assign workspaces to an org. Thin wrapper over updateOrg. */
export async function assignWorkspaces(ref: string, workspaces: string[]): Promise<OrgAssignResult> {
  const result = await updateOrg(ref, { workspaces });
  return result.assigned ?? {};
}

/**
 * Delete an org.
 *
 * By default its data is reassigned to the default org — nothing is lost, only
 * the grouping. Pass purge to delete the workspaces, assets, vulnerabilities and
 * runs along with it.
 */
export async function deleteOrg(ref: string, purge = false): Promise<void> {
  if (isDemoMode()) {
    throw new Error("400:Deleting orgs is disabled in demo mode");
  }
  await http.delete(`${API_PREFIX}/orgs/${encodeURIComponent(ref)}`, {
    params: purge ? { purge: true } : undefined,
  });
}

File: pkg/server/handlers/orgs.go
package handlers

import (
	"context"
	"errors"
	"strings"

	"github.com/gofiber/fiber/v2"
	"github.com/j3ssie/osmedeus/v5/internal/config"
	"github.com/j3ssie/osmedeus/v5/internal/database"
)

// ResolveOrgQuery reads the ?org= query parameter and resolves it to an org UUID.
//
// An absent or empty parameter returns an empty UUID, which every read path treats
// as "no org filter" — the response then spans all orgs, exactly as it did before
// orgs existed. An unknown org is an error rather than a silent empty result, so a
// typo does not look like "this org has no data".
//
// The second return value is a ready-to-return 400 response, so list handlers stay
// a two-line call. It is 400 rather than the 404 orgError would produce, because
// here the org is a malformed filter argument, not the resource being addressed.
func ResolveOrgQuery(ctx context.Context, c *fiber.Ctx) (string, error) {
	ref := strings.TrimSpace(c.Query("org"))
	if ref == "" {
		return "", nil
	}
	orgUUID, err := database.ResolveOrgUUID(ctx, ref)
	if err != nil {
		return "", c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error":   true,
			"message": err.Error(),
		})
	}
	return orgUUID, nil
}

// orgStatusFor maps an org error to an HTTP status.
func orgStatusFor(err error) int {
	switch {
	case errors.Is(err, database.ErrOrgNotFound):
		return fiber.StatusNotFound
	case errors.Is(err, database.ErrDefaultOrgImmutable):
		return fiber.StatusForbidden
	default:
		return fiber.StatusInternalServerError
	}
}

func orgError(c *fiber.Ctx, err error) error {
	return c.Status(orgStatusFor(err)).JSON(fiber.Map{
		"error":   true,
		"message": err.Error(),
	})
}

// ListOrgs handles listing all orgs
// @Summary List orgs
// @Description Get every org with its workspace, asset and vulnerability counts
// @Tags Orgs
// @Produce json
// @Success 200 {object} map[string]interface{} "List of orgs"
// @Failure 500 {object} map[string]interface{} "Failed to fetch orgs"
// @Security BearerAuth
// @Router /osm/api/orgs [get]
func ListOrgs(cfg *config.Config) fiber.Handler {
	return func(c *fiber.Ctx) error {
		ctx := context.Background()

		orgs, err := database.ListOrgs(ctx)
		if err != nil {
			return orgError(c, err)
		}

		data := make([]fiber.Map, 0, len(orgs))
		for i := range orgs {
			org := &orgs[i]
			stats, err := database.GetOrgStats(ctx, org.UUID)
			if err != nil {
				return orgError(c, err)
			}
			data = append(data, fiber.Map{
				"uuid":        org.UUID,
				"name":        org.Name,
				"description": org.Description,
				"tags":        org.Tags,
				"is_default":  org.IsDefault(),
				"created_at":  org.CreatedAt,
				"updated_at":  org.UpdatedAt,
				"stats":       stats,
			})
		}

		return c.JSON(fiber.Map{"data": data, "total": len(data)})
	}
}

// GetOrg handles fetching a single org
// @Summary Get org
// @Description Get one org by name or UUID
// @Tags Orgs
// @Produce json
// @Param uuid path string true "Org name or UUID"
// @Success 200 {object} map[string]interface{} "Org details"
// @Failure 404 {object} map[string]interface{} "Org not found"
// @Security BearerAuth
// @Router /osm/api/orgs/{uuid} [get]
func GetOrg(cfg *config.Config) fiber.Handler {
	return func(c *fiber.Ctx) error {
		ctx := context.Background()

		org, err := database.ResolveOrgRef(ctx, c.Params("uuid"))
		if err != nil {
			return orgError(c, err)
		}
		return c.JSON(fiber.Map{"data": org})
	}
}

// GetOrgStats handles fetching an org's aggregate counts
// @Summary Get org stats
// @Description Get workspace, asset, vulnerability and run counts for an org
// @Tags Orgs
// @Produce json
// @Param uuid path string true "Org name or UUID"
// @Success 200 {object} map[string]interface{} "Org statistics"
// @Failure 404 {object} map[string]interface{} "Org not found"
// @Security BearerAuth
// @Router /osm/api/orgs/{uuid}/stats [get]
func GetOrgStats(cfg *config.Config) fiber.Handler {
	return func(c *fiber.Ctx) error {
		ctx := context.Background()

		org, err := database.ResolveOrgRef(ctx, c.Params("uuid"))
		if err != nil {
			return orgError(c, err)
		}

		stats, err := database.GetOrgStats(ctx, org.UUID)
		if err != nil {
			return orgError(c, err)
		}
		return c.JSON(fiber.Map{"data": stats})
	}
}

// createOrgRequest is the body for POST /osm/api/orgs
type createOrgRequest struct {
	Name        string   `json:"name"`
	Description string   `json:"description"`
	UUID        string   `json:"uuid"`
	Tags        []string `json:"tags"`
}

// CreateOrg handles creating an org
// @Summary Create org
// @Description Create a new org
// @Tags Orgs
// @Accept json
// @Produce json
// @Param body body createOrgRequest true "Org to create"
// @Success 201 {object} map[string]interface{} "Created org"
// @Failure 400 {object} map[string]interface{} "Invalid request"
// @Security BearerAuth
// @Router /osm/api/orgs [post]
func CreateOrg(cfg *config.Config) fiber.Handler {
	return func(c *fiber.Ctx) error {
		var req createOrgRequest
		if err := c.BodyParser(&req); err != nil {
			return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
				"error":   true,
				"message": "invalid request body: " + err.Error(),
			})
		}
		if strings.TrimSpace(req.Name) == "" {
			return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
				"error":   true,
				"message": "name is required",
			})
		}

		ctx := context.Background()
		org, err := database.CreateOrg(ctx, req.Name, req.Description, req.UUID, req.Tags)
		if err != nil {
			return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
				"error":   true,
				"message": err.Error(),
			})
		}

		return c.Status(fiber.StatusCreated).JSON(fiber.Map{"data": org})
	}
}

// updateOrgRequest is the body for PUT /osm/api/orgs/:uuid
type updateOrgRequest struct {
	Name        string   `json:"name"`
	Description string   `json:"description"`
	Tags        []string `json:"tags"`
	// Workspaces, when non-empty, assigns those workspaces to this org and
	// cascades the stamp to their assets, vulnerabilities and runs.
	Workspaces []string `json:"workspaces"`
}

// UpdateOrg handles updating an org and optionally assigning workspaces to it
// @Summary Update org
// @Description Update an org's metadata, and optionally assign workspaces to it
// @Tags Orgs
// @Accept json
// @Produce json
// @Param uuid path string true "Org name or UUID"
// @Param body body updateOrgRequest true "Fields to update"
// @Success 200 {object} map[string]interface{} "Updated org"
// @Failure 403 {object} map[string]interface{} "Default org cannot be renamed"
// @Failure 404 {object} map[string]interface{} "Org not found"
// @Security BearerAuth
// @Router /osm/api/orgs/{uuid} [put]
func UpdateOrg(cfg *config.Config) fiber.Handler {
	return func(c *fiber.Ctx) error {
		var req updateOrgRequest
		if err := c.BodyParser(&req); err != nil {
			return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
				"error":   true,
				"message": "invalid request body: " + err.Error(),
			})
		}

		ctx := context.Background()

		org, err := database.ResolveOrgRef(ctx, c.Params("uuid"))
		if err != nil {
			return orgError(c, err)
		}

		updated, err := database.UpdateOrg(ctx, org.UUID, req.Name, req.Description, req.Tags)
		if err != nil {
			return orgError(c, err)
		}

		response := fiber.Map{"data": updated}

		if len(req.Workspaces) > 0 {
			counts, err := database.AssignWorkspacesToOrg(ctx, org.UUID, req.Workspaces)
			if err != nil {
				return orgError(c, err)
			}
			response["assigned"] = counts
		}

		return c.JSON(response)
	}
}

// DeleteOrg handles deleting an org
// @Summary Delete org
// @Description Delete an org. Its data moves to the default org unless purge=true.
// @Tags Orgs
// @Produce json
// @Param uuid path string true "Org name or UUID"
// @Param purge query bool false "Also delete the org's workspaces, assets, vulnerabilities and runs" default(false)
// @Success 200 {object} map[string]interface{} "Deletion result"
// @Failure 403 {object} map[string]interface{} "Default org cannot be deleted"
// @Failure 404 {object} map[string]interface{} "Org not found"
// @Security BearerAuth
// @Router /osm/api/orgs/{uuid} [delete]
func DeleteOrg(cfg *config.Config) fiber.Handler {
	return func(c *fiber.Ctx) error {
		ctx := context.Background()

		org, err := database.ResolveOrgRef(ctx, c.Params("uuid"))
		if err != nil {
			return orgError(c, err)
		}

		purge := c.QueryBool("purge", false)
		if err := database.DeleteOrg(ctx, org.UUID, purge); err != nil {
			return orgError(c, err)
		}

		return c.JSON(fiber.Map{
			"message": "org deleted",
			"uuid":    org.UUID,
			"name":    org.Name,
			"purged":  purge,
		})
	}
}
